### ISSUE TITLE
Make the "partial" of a QNode batch-executable

### ISSUE THREAD

**Issue by dwierichs** [PennyLaneAI/pennylane#1881]
### Feature details

It would be nice, for example in the context of function reconstruction, to be able to batch execute a univariate restriction of a QNode. Consider the following QNode and its three univariate restrictions:

```python
@qml.qnode(dev)
def circuit(x, Y):
    qml.RX(x, wires=0)
    qml.RY(Y[0], wires=0)
    qml.RY(Y[0], wires=1)
    return qml.expval(qml.PauliZ(0))
    
x = np.array(0.3, requires_grad=True)
Y = np.array([-0.6, 0.2], requires_grad=True)
univ_0 = functools.partial(circuit, Y=Y)
univ_1 = lambda y0: circuit(x, Y+np.eye(2)[0]*y0)
univ_2 = lambda y1: circuit(x, Y+np.eye(2)[1]*y1)
```
Then it is possible to batch execute `circuit` but neither `univ_0` nor `univ_1/2` via `batch_params`. 
As the evaluations during a reconstruction task may be independent in many cases (if there is no advanced on-the-fly shot allocation or so), this would be a great use case for batching such executions together.

### Implementation

Essentially, one could implement this by including the restriction to a univariate function within a batch transform, so that calling
```python
qml.univariate_batch(circuit, argnum={"x": 0}, shifts=[-np.pi/2, 0.0, np.pi/2])(x, Y)
```
under the hood would do something like
```python
batched_x = [x+shift for shift in shifts]
batch_args = (batched_x,) + tuple([arg]*len(shifts) for arg in args)
qml.batch_params(circuit)(batch_args)
```

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by dwierichs** [comment]
One implementation would be to copy `batch_params` and skip/modify certain validation while copying non-batched parameters of the tape over to all newly created tapes of the batch. It could look like this:

```python
@batch_transform
def batch_params_univariate(tape, all_operations=False):

    params = tape.get_parameters(trainable_only=not all_operations)
    batched_ids = []
    batched_params = []
    for par_idx, par in enumerate(params):
        if len(qml.math.shape(par))>0:
            batched_ids.append(par_idx)
            batched_params.append(par)
    if batched_ids == []:
        raise ValueError("Could not find a batch dimension for any parameter.")

    batch_dims = list(map(len, batched_params))
    if not all(b_dim==batch_dims[0] for b_dim in batch_dims[1:]):
        raise ValueError(f"Mismatch in batched dimensions: {batch_dims}")

    output_tapes = []
    for batch_idx in range(batch_dims[0]):
        new_params = [(par if par_idx not in batched_ids else par[batch_idx]) for par_idx, par in enumerate(params)]
        new_tape = tape.copy(copy_operations=True)
        new_tape.set_parameters(new_params, trainable_only=not all_operations)
        output_tapes.append(new_tape)

    return output_tapes, lambda x: qml.math.squeeze(qml.math.stack(x))
```

As an alternative, one could build a "qnode transform" only, skipping the `batch_transform` template. This could make sense because the meaning of "univariate restriction" of a QNode  does not copy over to a tape. It could look like this:

```python
def univariate_batch(qnode, argnum, shifts):
    arg_names = list(signature(qnode).parameters.keys())
    arg_idx = arg_names.index(argnum[0])
    batching_vector = qml.math.ones(len(shifts))
    batched_qnode = qml.batch_params(qnode)

    @wraps(qnode)
    def wrapper(*args, **kwargs):
        batched_arg = args[arg_idx]
        if len(qml.math.shape(batched_arg)) == 0:
            shift_vec = qml.math.ones_like(batched_arg)
        else:
            shift_vec = qml.math.zeros_like(batched_arg)
            shift_vec[argnum[1]] = 1.0
        batched_arg = qml.math.tensordot(batched_arg, batching_vector, axes=0)
        for i, _shift in enumerate(shifts):
            batched_arg[...,i] += shift_vec*_shift

        pre_args = tuple(qml.math.tensordot(arg, batching_vector, axes=0) for arg in args[:arg_idx])
        post_args = tuple(qml.math.tensordot(arg, batching_vector, axes=0) for arg in args[arg_idx+1:])
        return batched_qnode(*pre_args, batched_arg, *post_args)

    return wrapper
```

**Comment by josh146** [comment]
Nice solutions @dwierichs! I suppose it is important to work out what particular feature we want to enable.

1. Do we want to univariate batch over gate arguments?

2. Or, do we want to univariate batch over QNode arguments?

`@qml.batch_params` solves (1) (in a general sense, not in the univariate case), by first allowing the tape to be constructed with gate arguments with batch dimension. It then transforms `tape with batch dims -> [list of tapes per batch dim]`, and executes each resulting tape simultaneously.

For (2), we currently have no general solution for this in the code base. This has been on my mind for a while; it sounds like a general solution would look like this:

```
input QNode -> output QNode
```

where the output QNode contains _multiple_ tapes?

**Comment by dwierichs** [comment]
1) Agreed, although `qml.batch_params` requires all parameters to be batched with the same dimension, not allowing conveniently for univariate batching (Instead, one has to outer-product the constant arguments with `qml.math.ones` because of the input format that `qml.batch_params` is using). So even for univariate batching over gate arguments, there would be some convenience method in order, I feel.

2) Yes, I think a QNode transform that is _not_ built on top of a tape transform might be nice to have as a concept here. I'm not sure it needs its own class like `tape_transform`/`batch_transform`, but something like the second code snippet above could be such a transform.

For the current purpose we had in mind (Rotosolve, finite differences), I think that scenario 2) is more relevant.
